File: LeetCode/898. Bitwise ORs of Subarrays.py
class Solution:
    def subarrayBitwiseORs(self, arr: list[int]) -> int:
        res = set()
        prev = set()
        for i in arr:
            cur = set()
            cur.add(i)
            for j in prev:
                cur.add(j | i)
            for k in cur:
                res.add(k)
            prev = cur
        return len(res) 
x = Solution()
print(x.subarrayBitwiseORs([1,1,2]))  # 3
print(x.subarrayBitwiseORs([0]))    # 1
print(x.subarrayBitwiseORs([1,2,4])) # 6 

File: LeetCode/898. Bitwise ORs of Subarrays.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var subarrayBitwiseORs = function(arr) {
    let res = new Set()
    let prev = new Set()
    for(let i of arr){
        let cur = new  Set()
        cur.add(i)
        for(let j of prev){
            cur.add(j | i)
        }
        for(let k of cur){
            res.add(k)
        }
        prev = cur 
    }
    return res.size;
}
console.log(subarrayBitwiseORs([0])) // 1
console.log(subarrayBitwiseORs([1,1,2])) // 3
console.log(subarrayBitwiseORs([1,2,4])) // 6 
